### DISCUSSION TITLE
Initial condition for nodal variable on inter-block interface

### DISCUSSION THREAD

**Discussion by Wendy-Ji** [idaholab/moose#26572] [Q&A General]
Hello,

In the ICs System section, I saw that nodal variables could be defined on the interface between two blocks. I was wondering whether this worked for changing the mesh. 

I have been using SolutionUserObject and SolutionFunction to read in the final timestep from an existing exodus file to use as an initial condition for a different input file with a different mesh. If I remove blocks from my mesh for the second input file, the variables are read correctly, but if I add blocks to the mesh, I get this error:

```
*** ERROR ***
The following error occurred in the object "prev_uo", of type "SolutionUserObject".

Failed to access the data for variable 'dummy' at point (x,y,z)=( 1.04226,     -0.5,      0.5) in the 'prev_uo' SolutionUserObject
```
This point seems to be near to the boundary where the block is added. I'm not sure how to set this up correctly to prevent the error. I'm currently using block restrictions on the initial conditions to use the exodus file for the existing block, and to set an initial condition for the added block (see below input files). 

Also, I've realised this setup does work for elemental variables, so I think I've done something wrong for the nodal variables. 

Thank you.

```
[Problem]
  kernel_coverage_check = false 
[]

[Mesh]
  [load_mesh] 
    type = FileMeshGenerator 
    file = 'block_1_only.e' 
  []
[]

[Variables] 
  [dummy]
    family = LAGRANGE
    order = FIRST
    initial_condition = 1.0
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  end_time = 2
[]

[Outputs]
  exodus = true
[]
```

```
[Problem]
  kernel_coverage_check = false 
[]

[Mesh]
  [load_mesh] 
    type = FileMeshGenerator 
    file = 'block_1_2.e' 
  []
[]

[Variables] 
  [dummy]
    family = LAGRANGE
    order = FIRST
  []
[]

[ICs]
  [prev_dummy]
    type = FunctionIC
    variable = 'dummy'
    function = 'prev_func'
    block = 'block_1'
  []
  [init_dummy]
    type = ConstantIC
    variable = 'dummy'
    value = 2
    block = 'block_2'
  []
[]

[Functions]
  [prev_func]
    type = SolutionFunction
    from_variable = 'dummy'
    solution = 'prev_uo'
  []
[]

[UserObjects]
  [prev_uo]
    type = SolutionUserObject
    mesh = 'block_1_init_out.e'
    timestep = 'LATEST'
    system_variables = 'dummy'
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  end_time = 2
[]

[Outputs]
  exodus = true
[]
```

**Comment by GiudGiud** [comment]
Hello

The solution user object attempts to read the variable at the desired initial condition locations (depends on the variable type). 

Probably when the blocks are not added, the search ends near a valid value. When the blocks are added, the search ends outside of the valid values in the solution file. 

Can you please make sure all the region it tries to initialize from are within the solution mesh? Maybe block restrict the initial condition to make sure it sits within the solution file mesh extent

Guillaume

**Reply by Wendy-Ji** [reply]
Hello,

In my input file (in the original post), I have block restricted the initial conditions to 'block_1' and 'block_2', which are the only blocks in the new mesh (block 2 was added, block 1 was the original). The error refers to a point which would lie in block 2. Is there something I have restricted incorrectly? Thank you.
 
Wendy

**Reply by GiudGiud** [reply]
So how is it going to m ow the values in block 2 from the solution file if it was added only in the new mesh?

**Reply by Wendy-Ji** [reply]
Hello,

I have separated my initial condition into two sections using block restrictions. One is for the original block 1, which would have values from the solution file, and the other is for the added block 2, which does not have values in the solution file and is instead set to a constant value of 2. I thought this would set the added block to an initial value of 2, but it looks like it's still trying to read it from the solution file, which then gives the error. This did work for elemental variables, so I'm not sure if it has something to do with the nodal variable specifically.

Wendy

**Reply by GiudGiud** [reply]
I ll take a look at the end of the week. 

This doesn't look normal indeed. 

If you want to keep working with this for now, I would try with a SolutionAux executed on initial instead of a solutionIC.

If that does work, I would load the solution in a mesh with a single block in a multiapp. Then transfer the solution from this block in the multiapp to the relevant block in main solve, the one that has both blocks

**Reply by Wendy-Ji** [reply]
Hello,

The SolutionAux executed on initial works, although the variable is now an auxvariable. I'm not sure if that would have any differences to solving it as a variable.

I'm also not sure I understand the multiapp method you talked about. Is the main app the one with the added block, and the subapp the original mesh? Would the transfer then be something like MultiAppNearestNodeTransfer, and the main app executes after the subapp completes?

Wendy

**Reply by GiudGiud** [reply]
Ah you are right. SolutionAux only works on auxvariables

And if you plan to solve for that variable, it must be a nonlinear variable not an aux variable

Multiapp is the way to go for now. 

Create a simulation which only intializes the variable in one block. 
Then nest that simulation in your main simulation using a FullSolveMultiapp executed on initial. 

Then use the transfers to move the fields from the multiapp to the nonlinear variable

**Reply by Wendy-Ji** [reply]
Hello,

The multiapp setup works for me. I ended up using the MultiAppGeneralFieldNearestLocationTransfer, and the 'from_blocks' and 'to_blocks' parameters to restrict the initial condition to the right block. Thank you for the workaround.

Wendy
